File: problems/56-merge-intervals/sort.js
/**
 * 按left排序，然后遍历即可
 */
var merge = function(intervals) {
  if (!intervals.length) return []

  intervals = intervals.slice().sort((a, b) => a[0] - b[0])

  const res = [intervals[0]]
  for (const interval of intervals.slice(1)) {
    const last = res[res.length - 1]
    if (canJoin(last, interval)) {
      res.push(join(res.pop(), interval))
    } else {
      res.push(interval)
    }
  }
  return res
};

function canJoin (A, B) {
  return !(A[1] < B[0] || B[1] < A[0])
}

function join (A, B) {
  const left = Math.min(A[0], B[0])
  const right = Math.max(A[1], B[1])
  return [left, right]
}

[
  [[1,3],[2,6],[8,10],[15,18]],
  [[1,4],[4,5]],
].forEach(arr => {
  console.log(merge(arr))
})

File: problems/56-merge-intervals/sort.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        const int n = intervals.size();
        if (n == 0) return {};

        // 排序：start小的优先
        sort(intervals.begin(), intervals.end(), [&](vector<int> &A, vector<int> &B) {
            return A[0] < B[0];
        });

        vector<vector<int>> res;
        vector<int> prev = intervals[0];

        // 只需要检查相邻的区间
        for (int i = 1; i < n; ++i) {
            vector<int> curr = intervals[i];

            if (curr[0] <= prev[1]) { // 相交 => 合并，即扩展右边界
                prev[1] = max(prev[1], curr[1]);
            } else { // 不相交 => 将上一个合并完成的加入结果集，并重置
                res.push_back(prev);
                prev = curr;
            }
        }

        // 最后一个
        res.push_back(prev);

        return res;
    }
};